File: apps/android-native/app/src/main/java/com/alarmtalk/app/ui/main/PromoPromptStore.kt
package com.alarmtalk.app

import android.content.Context

/**
 * 웰컴 프로모 코드 안내를 이 계정에 이미 띄웠는지 기록한다.
 *
 * 계정 단위인 이유: 코드 등록은 계정에 붙는 혜택이라, 같은 기기에서 다른 계정으로 들어온
 * 사람에게는 다시 안내해야 한다. 반대로 이미 한 번 본 계정에는 다시 띄우지 않는다 —
 * 첫 실행에는 이미 동의·권한·목소리 준비가 줄지어 있어서, 여기에 반복 노출까지 얹으면
 * 안내가 아니라 조르기가 된다.
 *
 * 재설치하면(allowBackup=false) 초기화돼 다시 한 번 뜬다. 코드를 이미 쓴 사람은 유료
 * 플랜이 되어 노출 조건(무료 플랜)에서 빠지므로 실질적인 중복 노출은 아니다.
 */
internal class PromoPromptStore(context: Context) {
    private val prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    fun hasPrompted(userId: String): Boolean = prefs.getBoolean(key(userId), false)

    fun markPrompted(userId: String) {
        prefs.edit().putBoolean(key(userId), true).apply()
    }

    private fun key(userId: String) = "promo_prompted_$userId"

    private companion object {
        const val PREFS_NAME = "alarmtalk_promo_prompt"
    }
}

File: apps/ios-native/AlarmTalk/PromoPromptStore.swift
import Foundation

/// 웰컴 프로모 안내를 이 **계정**에 이미 띄웠는지 기록한다.
///
/// 안드로이드 `ui/main/PromoPromptStore.kt`.
/// 계정 단위인 이유: 코드 등록은 계정에 붙는 혜택이라, 같은 기기에서 다른 계정으로
/// 들어온 사람에게는 다시 안내해야 한다. 반대로 한 번 본 계정에는 다시 띄우지 않는다 —
/// 첫 실행에는 이미 동의·권한·목소리 준비가 줄지어 있어, 반복 노출까지 얹으면 안내가
/// 아니라 조르기가 된다.
struct PromoPromptStore {
    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) { self.defaults = defaults }

    func hasPrompted(userID: String) -> Bool { defaults.bool(forKey: key(userID)) }
    func markPrompted(userID: String) { defaults.set(true, forKey: key(userID)) }

    private func key(_ userID: String) -> String { "promo_prompted_\(userID)" }
}
